### PR TITLE
V2 API implementation for token generate/refresh

### PR DESCRIPTION
**Pull request by haitaottd**

### CODE REVIEW

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java [hunk 1]
```diff
@@ -230,35 +273,52 @@ private void handleHealthCheck(RoutingContext rc) {
...
+            ResponseUtil.Error(ResponseStatus.UnknownError, 500, rc, "Service Error");
+        }
+    }
+
+    private void handleRefreshFailure(RoutingContext rc, RefreshResponse r) {
+        if (r.isOptOut() || r.isDeprecated()) {
+            ResponseUtil.Error(ResponseStatus.OptOut, 400, rc, "User opt out");
+        } else if (!AuthMiddleware.isAuthenticated(rc)) {
+            // unauthenticated clients get a generic error
+            ResponseUtil.Error(ResponseStatus.GenericError, 400, rc, "Error refreshing token");
```
Comment (scong-ttd) [inline src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:314]: why not 401 here?

File: src/main/java/com/uid2/operator/service/V2EncryptedTokenEncoder.java [hunk 2]
```diff
@@ -37,16 +37,19 @@ public class V2EncryptedTokenEncoder implements ITokenEncoder {
     private final IKeyStore keyStore;
     private final SecureRandom random = new SecureRandom();
 
-    public V2EncryptedTokenEncoder(IKeyStore keyStore) {
+    private final boolean useGCM;
```
[the remark this answers is not in the document]
Comment (haitaottd) [reply src/main/java/com/uid2/operator/service/V2EncryptedTokenEncoder.java:40]: It's recommended by security team

File: src/main/java/com/uid2/operator/service/V2EncryptedTokenEncoder.java [hunk 3]
```diff
@@ -209,6 +220,9 @@ private byte[] encryptIdentity(UserIdentity identity, EncryptionKey key) {
         }
         b.appendInt(identity.getPrivacyBits());
         b.appendLong(identity.getEstablished().toEpochMilli());
-        return EncryptionHelper.encrypt(b.getBytes(), key).getPayload();
+        return useGCM ?
```
[the remark this answers is not in the document]
Comment (haitaottd) [reply src/main/java/com/uid2/operator/service/V2EncryptedTokenEncoder.java:223]: See above

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java (hunk 1 above)
Comment (haitaottd) [reply src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:314]: Not sure. It's like this [before](https://github.com/IABTechLab/uid2-operator/blob/master/src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java#:~:text=unauthenticated%20clients%20get%20a%20generic%20error). But this is obsolete with V2 since V2 refresh endpoint requires client key, while V1 doesn't.

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java [hunk 4]
```diff
@@ -265,11 +270,18 @@ private void handleTokenRefreshV1(RoutingContext rc) {
...
+            refreshToken = v2Json.getString("refresh_token");
+        }
+
         try {
-            RefreshResponse r = idService.refreshIdentity(tokenList.get(0));
+            RefreshResponse r = idService.refreshIdentity(refreshToken);
             if (!r.isRefreshed()) {
                 if (r.isOptOut() || r.isDeprecated()) {
-                    ResponseUtil.SuccessNoBody(ResponseStatus.OptOut, rc);
+                    ResponseUtil.Error(ResponseStatus.OptOut, 400, rc, "User opt out");
```
Comment (tqinli) [inline src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:284]: breaking change for v1?

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java [hunk 5]
```diff
@@ -287,7 +299,32 @@ private void handleTokenRefreshV1(RoutingContext rc) {
...
+                } else if (r.isExpired()) {
+                    ResponseUtil.Error(ResponseStatus.ExpiredToken, 400, rc, "Expired Token presented");
+                } else {
+                    ResponseUtil.Error(ResponseStatus.UnknownError, 500, rc, "Unknown State");
+                }
+            } else {
+                ResponseUtil.SuccessV2(rc, toJsonV1(r.getTokens()));
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (tqinli) [inline src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:325]: LOGGER.error(e)

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java [hunk 6]
```diff
@@ -287,7 +299,32 @@ private void handleTokenRefreshV1(RoutingContext rc) {
             e.printStackTrace();
```
Comment (tqinli) [inline src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:299]: fix here too :-)

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java (hunk 4 above)
Comment (haitaottd) [reply src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:284]: Yes. I think we should only say success if token is refreshed. And it's easier for client to handle. If it's http 200, token is refreshed and response needs decryption. Otherwise, no decryption and check failure reason in plain response json.

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java (hunk 6 above)
Comment (haitaottd) [reply src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:299]: done. Replace all printStackTrace() with LOGGER.error(e)

File: src/main/java/com/uid2/operator/service/V2EncryptedTokenEncoder.java [hunk 7]
```diff
@@ -58,7 +61,9 @@ public byte[] encode(AdvertisingToken t) {
         b2.appendLong(t.getExpiresAt().toEpochMilli());
         encodeSiteIdentity(b2, t.getIdentity(), siteEncryptionKey);
 
-        byte[] encryptedId = EncryptionHelper.encrypt(b2.getBytes(), masterKey).getPayload();
+        byte[] encryptedId = useGCM ?
```
[the remark this answers is not in the document]
Comment (haitaottd) [reply src/main/java/com/uid2/operator/service/V2EncryptedTokenEncoder.java]: This is reverted. Token internal encryption still uses CBC.

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java (hunk 4 above)
Comment (atarassov-ttd) [reply src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:284]: Optout is not an error. Please revert back to the original behavior.

Comment (haitaottd) [reply src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java]: restored v1 behavior

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java [hunk 8]
```diff
@@ -333,12 +376,12 @@ private void handleTokenValidateV2(RoutingContext rc) {
...
                 return;
             }
-            if (ValidationInput.equals(input.getIdentityInput())) {
+            if (Arrays.equals(ValidationInput, input.getIdentityInput())) {
                 try {
                     final Instant now = Instant.now();
                     final String token = req.getString("token");
 
-                    if (this.idService.advertisingTokenMatches(token, input.toUserIdentity(IdentityScope.UID2, 0, now), now)) {
+                    if (this.idService.advertisingTokenMatches(token, input.toUserIdentity(IdentityScope.UID2, 0, now), Instant.now())) {
```
Comment (atarassov-ttd) [inline src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:384]: What is the reason for this change?

File: src/main/java/com/uid2/operator/service/V2RequestUtil.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    public static final int V2_REFRESH_PAYLOAD_LENGTH = 364;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(V2RequestUtil.class);
+
+    public static V2Request parseRequest(String bodyString, ClientKey ck) {
+        // Payload envelop format:
+        //  byte 0: version
+        //  byte 1-12: GCM IV
+        //  byte 13-end: encrypted payload + GCM AUTH TAG
+        byte[] bodyBytes = Utils.decodeBase64String(bodyString);
```
Comment (atarassov-ttd) [inline src/main/java/com/uid2/operator/service/V2RequestUtil.java:58]: Do I understand correctly that if this or logic below throws, we'd respond with 500 to client? Can we avoid that?

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java (hunk 8 above)
Comment (haitaottd) [reply src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:384]: restored. Probably missed during merge

File: src/main/java/com/uid2/operator/service/V2RequestUtil.java (hunk 9 above)
Comment (haitaottd) [reply src/main/java/com/uid2/operator/service/V2RequestUtil.java:58]: Added try...catch for base64 decode exception.

**Comment by atarassov-ttd** [comment]
LGTM

**Review by atarassov-ttd — APPROVED** [review]
